### ISSUE TITLE
Stray socket connections when sendning Emails

### ISSUE THREAD

**Issue by bkolicoski** [googleapis/google-api-php-client#2227]
I'm using the api client to send out scheduled emails through Gmail and that all works as expected. However, since the code runs through the Laravel scheduler, the code is kept in memory at all times and every sent email causes a connection to stay open. 

After a while, the server reaches the "Maximum open files" limit and further processing fails because of it. At first I though that it might be an issue with my code and added unset explicitly to any references to gmail objects but the issue still remains. 

To track it down, i used the following command to list out the open file count and the process IDs. 

```
lsof | awk '{print $2}' | sort | uniq -c | sort -n
```

This returns a list that looks like the one below:
![image](https://user-images.githubusercontent.com/9977804/158577144-19e0eda7-cede-4954-acb9-8bbb48ace0fa.png)

Next, with the PID of the troubling process, which is the same as the process processing the Laravel queue I'm running the following command to see the open files of it. (590981 is the current process ID)
```
ls -l /proc/590981/fd
```
And that shows that all of the open files are actually socket connections. 
![image](https://user-images.githubusercontent.com/9977804/158577502-8f3a622b-32d0-41c7-89a8-e4e981d26d30.png)

To explore further what those connections are, I used the command below to get a list of open connections for that process:
```
lsof -i -a -p 590981
```
![image](https://user-images.githubusercontent.com/9977804/158578063-8cc9d9d4-0f22-4150-b9ea-d7e4b0d9312d.png)

As visible on the image, all those open socket connections are made to the `1e100.net` domain that is owned by Google, and from my understanding, it is used for tracking the requests. 

I did not find any direct reference to it in the code, nor a way how I can properly close this connection after the email is sent. 

#### Environment details

  - OS: Ubuntu 20.04
  - PHP version: 8.1
  - Package name and version: google/apiclient - v2.12.1 

#### Code example

```php
$client = new Client();
$client->setAuthConfig(base_path('/') . config('services.google.service_file'));
$client->addScope([
  Gmail::GMAIL_SEND,
  Gmail::GMAIL_READONLY,
  Gmail::GMAIL_MODIFY
]);
$client->setRedirectUri(url('/api/emails/callback'));
$client->setAccessType('offline');
$client->setPrompt("consent");
$client->setIncludeGrantedScopes(true);

$service = new Gmail($client);

//$this->entry is obtained from the created job

$message = (new Swift_Message($this->getMessageSubject($this->entry['sending_message'], $this->entry['type'])))
  ->setFrom($this->entry['email']['email'], $this->entry['email']['name'])
  ->setTo($this->entry['sending_message']['email'])
  ->setContentType('text/html')
  ->setCharset('utf-8')
  ->setBody($this->getMessageBody($this->entry['sending_message'], $this->entry['type']));

//other code

$msg_base64 = base64_encode($message->toString());
unset($message);

$message = new Message();
$message->setRaw($msg_base64);
if ($this->entry['sending_message']['google_thread_id'] != null) {
  $message->setThreadId($this->entry['sending_message']['google_thread_id']);
}
$message = $service->users_messages->send('me', $message);

//other code to save the sent state

unset($message);
unset($service);
unset($client);
```

I found several other instances where people had the similar issue with the `1e100.net` domain, [here](https://github.com/googleapis/google-cloud-python/issues/5523), [here](https://github.com/googleapis/google-cloud-dotnet/issues/3121) and [here](https://stackoverflow.com/questions/34532951/google-compute-engine-socket-issue) but none of them are for PHP nor Gmail.

I'm basically clueless as to what might open these connections and how to have them properly closed. Any ideas are more than welcome. 

Thanks!

**Comment by bshaffer** [comment]
Hello @bkolicoski 

There is a call to `file_get_contents` in the API client to [read your credentials](https://github.com/googleapis/google-api-php-client/blob/main/src/Client.php#L983). You could try passing in a simple [cache object](https://github.com/googleapis/google-api-php-client/blob/main/src/Client.php#L1084), setting the credentials manually, or passing in a [stream context](https://stackoverflow.com/questions/34886386/php-file-get-contents-when-does-connection-close) to see if this is where the issue is happening - I am not sure if this could be causing your issue or not.

**Comment by bkolicoski** [comment]
OK @bshaffer , I've added the `stream_context` as suggested but it seems that it has no difference whatsoever on the number of created connections. 

I thought that the default logger might be the culprit and I replaced it with the built-in logger in Laravel but that did not make a difference as well. 

I'm convinced that the line causing the issue is:
```
$message = $service->users_messages->send('me', $message);
```
but I have no way of proving that and I've been exploring the source code for hours to try and understand where and how it might happen. I have other background jobs that read and check emails and they do not have this issue. Only the job that sends the emails is showing this issue. 

For the time being, I have a workaround to reset the worker every few hours in order to prevent it from ever reaching the limit but I would rather have the connections closed without reloading the worker.
